### PR TITLE
Buildable wall light

### PR DESCRIPTION
**Pull request by Macoron**
- [x] You can build and deconstruct  light fixture
- [x] Both bulbs and tube lights works
- [x] You can build light only on walls
- [x] You can build lights only facing right side

Known issues:
- ~~Bulb wall lights sprites were broken. The direction of sprites doesn't match with direction of entities. I fixed sprites, but maps already contains bulbs with wrong rotation.~~ Fixed
- You can build light on top of existing light

### CODE REVIEW

**Comment by Macoron** [comment]
Seems done, ready for feedback. However the biggest issue is that orientation of bulbs is broken on all maps.

Sprite for bulbs weren't correctly rotated, so all maps contains bulbs with wrong direction. The order of rotation was different from other lights sprites. Check screenshot (left - was, right - now):
image

Is it possible to automatically change object rotation on all maps?

**Comment by SweptWasTaken** [comment]
> 
> 
> Seems done, ready for feedback. However the biggest issue is that orientation of bulbs is broken on all maps.
> 
> Sprite for bulbs weren't correctly rotated, so all maps contains bulbs with wrong direction. The order of rotation was different from other lights sprites. Check screenshot (left - was, right - now):
> image
> 
> Is it possible to automatically change object rotation on all maps?

You could go into the map yml and manually edit the rotation value of all the lights

**Comment by Macoron** [comment]
I wrote the worst code in my life that looks like this:
```
        elif str == '    rot: -1.5707963267948966 rad\n':
            content[ptr] = '    rot: 1.5707963267948966 rad\n'
        elif str == '    rot: 1.5707963267948966 rad\n':
            content[ptr] = '    rot: -1.5707963267948966 rad\n'
        elif str == "    rot: 3.141592653589793 rad\n":
            content[ptr] = '    rot: 0 rad\n'
```

But it did the trick, so all lamps on all maps have correct orientation. Yay!

File: Resources/Prototypes/Recipes/Construction/Graphs/lighting.yml [hunk 1]
```diff
@@ -0,0 +1,46 @@
+﻿- type: constructionGraph
```
Comment (metalgearsloth) [inline Resources/Prototypes/Recipes/Construction/Graphs/lighting.yml:1]: @Zumorica do you mind me pinging you whenever construction comes up in a review?

File: Content.Shared/Construction/ConstructionConditions/WallmountCondition.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            // check that player doesn't try to build object facing wall other side
+            bool invalidDirection = FacingWallOtherSide(user, location, direction);
+            if (invalidDirection)
+                return false;
+
+            if (direction != Direction.North)
+            {
+                // check that building inside wall
+                var wallEnity = location.GetEntitiesInTile(true)
+                    .FirstOrDefault((e) => e.HasComponent<SharedWallComponent>());
```
Comment (metalgearsloth) [inline Content.Shared/Construction/ConstructionConditions/WallmountCondition.cs:29]: I'm really not a fan of adding SharedWall and think you should really just raycast instead probably, or just check collision properties, anything else.

When we get a bigger map (the end goal is roughly 8x - 10x the size of saltern) you're talking about potentially 10,000 - 20,000 dummy components just for this and pjb would probably shank me if I merged it.

Plus it doesn't handle other tile-sized blockers like vending machines, medical scanners, AME parts, etc.

File: Content.Shared/Construction/ConstructionConditions/WallmountCondition.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+                    return false;
+            }
+            else
+            {
+                // north is exception because of world projection
+                // wallmount wil be build one tile above the wall
+                var belowLocation = location.Offset(Direction.South);
+                var wallEnity = belowLocation.GetEntitiesInTile(true)
+                    .FirstOrDefault((e) => e.HasComponent<SharedWallComponent>());
+                if (wallEnity == null)
```
Comment (metalgearsloth) [inline Content.Shared/Construction/ConstructionConditions/WallmountCondition.cs:40]:
```suggestion
                var wallEntity = belowLocation.GetEntitiesInTile(true)
                    .FirstOrDefault((e) => e.HasComponent<SharedWallComponent>());
                if (wallEntity == null)
```

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
I assume the new pngs are from the same commits as well for licence reasons.

I also didn't realise lights use sprite netsync so I'm gonna make an issue to fix that.

File: Content.Shared/Construction/ConstructionConditions/WallmountCondition.cs (hunk 2 above)
Comment (RemieRichards) [reply Content.Shared/Construction/ConstructionConditions/WallmountCondition.cs:29]: This is why we need tags

Comment (metalgearsloth) [reply Content.Shared/Construction/ConstructionConditions/WallmountCondition.cs:29]: NPCs definitely need tags for certain things at some point but in this case couldn't you just raycast to the nearest edge, so if you're building on the farside it'll block on the wall but on the nearside it won't for example? Or am I underthinking it.

**Comment by SweptWasTaken** [comment]
Fix conflicts if you want this merged.

File: Content.Shared/Construction/ConstructionConditions/WallmountCondition.cs (hunk 2 above)
Comment (Macoron) [reply Content.Shared/Construction/ConstructionConditions/WallmountCondition.cs:29]: I need someway to check that this player trying build light on the wall. To distinguish wall from low wall or window, I need tag. Same logic used for low walls in [SharedCanBuildWindowOnTopComponent](https://github.com/space-wizards/space-station-14/blob/745401a41e219b88d2e71983c8f6b3a1aaa486e1/Content.Shared/Construction/ConstructionConditions/LowWallInTile.cs#L23)

So I'm not sure how raycast can fix that. Does SS14 some kind of raycast groups or layers? Should I add all walls to this group?

Comment (metalgearsloth) [reply Content.Shared/Construction/ConstructionConditions/WallmountCondition.cs:29]: Walls have a specific collision mask that other things don't have i.e. Opaque and MobImpassable together (not many things have Opaque) so you could probably check on that.

The raycast also means it'll handle building to the specific edge as well.

Comment (Macoron) [reply Content.Shared/Construction/ConstructionConditions/WallmountCondition.cs:29]: I'm afraid it won't work. There are about 146 prototypes that are Opaque and MobImpassable. Furniture, airlocks and so on. 
Should I add new collision group for walls?

Comment (metalgearsloth) [reply Content.Shared/Construction/ConstructionConditions/WallmountCondition.cs:29]: What about walls 5 layers?

**Comment by Macoron** [comment]
@metalgearsloth I moved wall check to raycast and separate 'Wall' layer. Seems to work fine.

File: Content.Shared/Physics/CollisionGroup.cs [hunk 4]
```diff
@@ -24,6 +24,7 @@ public enum CollisionGroup
         Underplating    = 1 <<  7, // 128 Things that are under plating
         Passable        = 1 <<  8, // 256 Things that are passable
         ExplosivePassable = 1 << 9, // 512 Things that let the pressure of a explosion through
+        Walls = 1 << 10, // 1024 Special group to "tag" solid walls. Used to check if player can build wallmount on it.
```
Comment (metalgearsloth) [inline Content.Shared/Physics/CollisionGroup.cs:27]: I'm always suss on adding another layer to the collision group especially for 1 thing (like Passable and ExplosivePassable are). Maybe just do the raycast and pass in a predicate for the wall light tag? @RemieRichards 2nd opinion plz.

Comment (Macoron) [reply Content.Shared/Physics/CollisionGroup.cs:27]: So, what should I do? Return back to SharedWallComponent?

Comment (metalgearsloth) [reply Content.Shared/Physics/CollisionGroup.cs:27]: Tags ended up getting merged so go with a tag. Probably raycast and use a predicate on the tag.

The only time collision layers should be added is if they're actually used for collision and not just raycasts.

Comment (Macoron) [reply Content.Shared/Physics/CollisionGroup.cs:27]: Sadly, tags are server-side right now, so I can't use them in shared namespace for build conditional.
Though after a short discussion tag component will be moved to shared namespace soon.

Comment (metalgearsloth) [reply Content.Shared/Physics/CollisionGroup.cs:27]: Should be good now as it was merged.

**Comment by Macoron** [comment]
Moved from collision layers to tags. Hope it's finally over.

File: Resources/Prototypes/Entities/Constructible/Walls/lighting.yml [hunk 5]
```diff
@@ -47,10 +50,38 @@
...
         acts: ["Destruction"]
 
+- type: entity
+  id: PoweredlightEmpty
+  parent: Poweredlight
+  components:
+  - type: Sprite
+    state: empty
+  - type: PoweredLight
+    hasLampOnSpawn: False
```
Comment (Zumorica) [inline Resources/Prototypes/Entities/Constructible/Walls/lighting.yml:60]: Add a suffix to this prototype specifying that it's empty

File: Resources/Prototypes/Entities/Constructible/Walls/lighting.yml [hunk 6]
```diff
@@ -79,24 +110,10 @@
         acts: ["Destruction"]
 
 - type: entity
-  name: unpowered small light
-  id: SmallLight
-  parent: WallLight
+  id: PoweredSmallLightEmpty
+  parent: PoweredSmallLight
```
Comment (Zumorica) [inline Resources/Prototypes/Entities/Constructible/Walls/lighting.yml:114]: Add a suffix to this prototype specifying that it's empty

**Review by Zumorica — APPROVED** [review]
LGTM, thank you! This can be merged after you address the small nitpicks

**Comment by SweptWasTaken** [comment]
@Zumorica Could you give this another review? Looks like he fixed the nitpicks.

**Comment by Zumorica** [comment]
> @Zumorica Could you give this another review? Looks like he fixed the nitpicks.

if he did, it's good for merge

**Pull request merged**